BUG: Index._searchsorted_monotonic(..., side='right') returns the left side position for monotonic decreasing indexes

jschendel opened the pull request:
- [X] closes #17271
- [X] closes #16417
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

I didn't add a whatsnew entry for the `Index._searchsorted_monotonic` fix, since it looks like past convention has been to not add whatsnew entries for private methods.  Happy to add a whatsnew entry if I'm mistaken though.  I did add a whatsnew entry for the downstream effect on IntervalIndex.

jschendel committed:
FIX: Index._searchsorted_monotonic for decreasing indexes

Fixed an issue where Index._searchsorted_monotonic(..., side='right')
returns the left side position for monotonic decreasing indexes.  Issue had a downstream effect on scalar lookups in IntervalIndex as well.

gfyoung commented:
> since it looks like past convention has been to not add whatsnew entries for private methods.

That is correct.

jreback commented on pandas/tests/indexes/test_base.py:
```diff
@@ -2103,3 +2103,24 @@ def test_intersect_str_dates(self):
+    def test_searchsorted_monotonic(self):
+        # GH17271
+        idx_inc = Index([0, 2, 4])
+        idx_dec = Index([4, 2, 0])
```
would be better to move these to ``common.py/Base`` as they be tested generically.

jreback commented on pandas/tests/indexing/test_interval.py:
```diff
@@ -39,6 +42,9 @@ def test_loc_with_scalar(self):
         expected = s.iloc[2:5]
         tm.assert_series_equal(expected, s.loc[s >= 2])
 
+        # test endpoint of non-overlapping monotonic decreasing (GH16417)
```
I would instead make a specific test of this (also test increasing monotonic), even better would be to parametrize over closed (bonus points)

jschendel committed:
Review related changes

jschendel commented:
Notes regarding the review changes I made:

- Removed some redundant checks from `test_loc_with_scalar` and `test_getitem_with_scalar` in `test_interval.py`, as these checks are now being handled by `test_nonoverlapping_monotonic`  (the index used in the two former tests is nonoverlapping monotonic increasing).

- Added some monotonic decreasing indexes to `setup_method` to various test classes within in order to ensure that monotonic decreasing indexes are fully tested by `common.py/test_searchsorted_monotonic` (and other tests).
   - Had to change ` tm.assert_numpy_array_equal` to `tm.equalContents` in `common.py/test_difference_base`.  Introducing a decreasing sequence led to the test failing with ` tm.assert_numpy_array_equal` due to different sorting of the result and expected answer.  If keeping ` tm.assert_numpy_array_equal` is preferable, the arrays could be sorted prior to using it.
   -  Did not add a monotonic decreasing index to `test_range`, as it currently fails `common.py/test_intersection_base`  due to the issue described in #17296 (will add when I submit a PR for this issue).

jreback commented:
> Had to change tm.assert_numpy_array_equal to tm.equalContents in common.py/test_difference_base. Introducing a decreasing sequence led to the test failing with tm.assert_numpy_array_equal due to different sorting of the result and expected answer. If keeping tm.assert_numpy_array_equal is preferable, the arrays could be sorted prior to using it.

yes, pls just sort them if need; we don't use ``tm.equalContents`` (actually should remove that)


> Did not add a monotonic decreasing index to test_range, as it currently fails common.py/test_intersection_base due to the issue described in #17296 (will add when I submit a PR for this issue).

simply add an xfailing test (with the issue number).

jschendel committed:
Additional review changes

jschendel commented:
Okay, made those additional changes.  Any idea why codecov is saying this will decrease coverage by 50%?  I don't think I made any substantial changes that would cause that.  Just a glitch?

jschendel commented:
ping @jreback : made review related updates and did a rebase to resolve merge conflicts

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -354,6 +354,7 @@ Indexing
 - Avoids ``IndexError`` when passing an Index or Series to ``.iloc`` with older numpy (:issue:`17193`)
 - Allow unicode empty strings as placeholders in multilevel columns in Python 2 (:issue:`17099`)
 - Bug in ``.iloc`` when used with inplace addition or assignment and an int indexer on a ``MultiIndex`` causing the wrong indexes to be read from and written to (:issue:`17148`)
+- Bug in ``IntervalIndex`` where performing a scalar lookup fails for included right endpoints of non-overlapping monotonic decreasing indexes (:issue:`16417`)
```
can you add both issue numbers

jreback commented on pandas/tests/indexes/common.py:
```diff
@@ -954,3 +957,24 @@ def test_join_self_unique(self, how):
+            value = index[0]
+
+            if index.is_monotonic_increasing or index.is_monotonic_decreasing:
+                assert index._searchsorted_monotonic(value, side='left') == 0
```
can you test ``.searchsorted()`` as well for the various cases

jreback commented on pandas/tests/indexing/test_interval.py:
```diff
@@ -67,6 +48,41 @@ def test_getitem_with_scalar(self):
 
+    @pytest.mark.parametrize('direction, closed',
+                             product(('increasing', 'decreasing'),
+                                     ('left', 'right', 'neither', 'both')))
```
@zfrenchee if you'd have a look here

jschendel committed:
Test searchsorted and misc changes

jschendel commented:
Made the requested changes.  A couple comments:

- Only tested `.searchsorted` for monotonic increasing, since that's the underlying assumption.
- Added an index with repeats to `test_base` to better test `side='right'` for `_searchsorted_monotonic` and `.searchsorted`.
   - Required a modifying `test_get_indexer_consistency` to check uniqueness in one place where uniqueness was assumed.
- Moved the xfail for `test_intersection_base` to only occur for `RangeIndex`, since that's the only type of index that should fail.

jreback commented on pandas/tests/indexes/common.py:
```diff
@@ -138,9 +138,10 @@ def test_get_indexer_consistency(self):
+            if index.is_unique:
+                indexer = index.get_indexer(index[0:2])
+                assert isinstance(indexer, np.ndarray)
+                assert indexer.dtype == np.intp
```
can you add what the result if not index.is_unique as well

jreback reviewed (changes requested):
minor request. lgtm. ping on green.

jschendel committed:
Verify non-unique index

jschendel commented:
ping @jreback : checks are green

Made one small additional change to the `index.is_unique` case.  Since `get_indexer` is implemented slightly differently for `CategoricalIndex` relative to other indexes, they're actually fine even if the values in the index aren't unique, so they fell into the `index.is_unique` case regardless of uniqueness.

jschendel commented:
ping @jreback : green again after rebasing to address the merge conflict.

jschendel committed:
Rebase and remove xfail

Removed xfail, as the issue impacting the failing test has been resolved.  Rebase to address merge conflicts.

jreback commented:
thanks @jschendel nice fixes! keep em coming!

jbrockmendel commented on pandas/tests/indexes/common.py:
```diff
@@ -138,9 +139,14 @@ def test_get_indexer_consistency(self):
-            indexer = index.get_indexer(index[0:2])
-            assert isinstance(indexer, np.ndarray)
-            assert indexer.dtype == np.intp
+            if index.is_unique or isinstance(index, CategoricalIndex):
```
@jschendel why is CategoricalIndex excluded here?
